Xamarin.PropertyEditing.Mac/Controls/Custom/HistoryLayer.cs

```diff
@@ -78,7 +78,11 @@ public override void UpdateFromModel (EditorInteraction interaction)
 		{
 			LayoutIfNeeded ();
 			current.BackgroundColor = interaction.Color.ToCGColor ();
-			previous.BackgroundColor = interaction.ViewModel.InitialColor.ToCGColor ();
+			//there are some scenarios where viewmodel could be null
+			var initialColor = interaction.ViewModel?.InitialColor;
```

Comment:
The problem is that these CGColor instances are not stored as fields.

Reply:
You need fields for these colors, xamarin.mac only sets the native pointer here, and they could go away.

Reply:
@Therzok - Are you saying that if CALayer.BackgroundColor is set to some CGColor, that alone won't keep it from getting GC'ed so you need to have another reference to that same color stored somewhere to prevent the GC? Weird - that seems like an easy mistake to make if true.

Reply:
So, there are different things at play here. NSColor used lifetime handling using NSObject wrappers with a ToggleRef. You can see that via native property annotations, like `strong`. In some cases, Xamarin.Mac is overzealous and adds `strong` managed fields for properties held `weak` by native.

CGColor uses old-school CFRetain/CFRelease lifetimes. Native holds onto a ref for the object, so something here doesn't really make sense..

> The value of this property is retained using the Core Foundation retain/release semantics. This behavior occurs despite the fact that the property declaration appears to use the default assign semantics for object retention.

```
QuartzCore!CA::AttrList::set(unsigned int, _CAValueType, void const*)
QuartzCore!CA::Layer::setter(unsigned int, _CAValueType, void const*)
QuartzCore!-[CALayer setBackgroundColor:]
VisualStudio!
Xamarin.Mac.dll!CoreAnimation.CALayer::set_BackgroundColor
Xamarin.PropertyEditing.Mac.dll!Xamarin.PropertyEditing.Mac.HistoryLay
```

Why would it crash in this code? If anything in here _were_ null, we'd crash in ToCGColor if anything was null here before setting it. So, I think the actual fix is to use something like `GC.KeepAlive` or store it in a field to avoid it getting GC'd while the PInvoke is happening.

Reply:
In short: GC ran and collected the CGColor before native retained it.

Reply:
> Weird - that seems like an easy mistake to make if true.

You're dealing with code that's almost identical to a plain PInvoke :p

Reply:
@Therzok - Ouch. And thanks. I would have hoped that Xamarin.Mac would have had some code to keep objects alive when they are passed along in PInvoke, but apparently not. Is there a pattern that's used in other places to address this? I was thinking I'd create a helper method to set CALayer.BackgroundColor, with the GC.KeepAlive, but if there's an existing pattern I could use that instead.

Reply:
No need to add _more_ APIs. The problem only exhibits with CoreFoundation types, not with NSObject types.

Reply:
@Therzok - I made an update to use GC.KeepAlive. Can you take a look at the latest change. As there are three places BackgroundColor is set, abstracting it to a separate method seemed to make sense.

Reply:
I opened a bug against Xamarin.Mac, so let's see if they can do anything in the code-generation layer: https://github.com/xamarin/xamarin-macios/issues/15600